src/WebformSubmissionLogMailer.php

```diff
@@ -0,0 +1,138 @@
+
+  /**
+   * Mailer constructor.
+   *
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $configFactory
+   *   The config factory service interface.
+   * @param \Drupal\Core\Mail\MailManager $mailManager
+   *   The mail manager service.
+   * @param \Drupal\Core\Language\LanguageManager $languageManager
+   *   The language manager service.
+   * @param \Drupal\Core\Routing\UrlGeneratorInterface $urlGenerator
+   *   The URL generator.
+   */
+  public function __construct(ConfigFactoryInterface $configFactory, MailManager $mailManager, LanguageManager $languageManager, UrlGeneratorInterface $urlGenerator) {
+    $this->configFactory = $configFactory;
+    $this->mailManager = $mailManager;
+    $this->languageManager = $languageManager;
+    $this->urlGenerator = $urlGenerator;
+  }
+
+  /**
+   * Send "failed job" mail to recipients related to webform.
+   *
+   * @param $webformSubmission
+   *   The webform submission that failed.
+   * @param $context
+   *   The logging context.
+   * @return void
+   */
+  public function sendMails($webformSubmission, $context) {
+    $webform = $webformSubmission->getWebform();
+    $webformSettings = $webform->getThirdPartySetting('os2forms', 'os2forms_webform_submission_log');
+
+    $recipients = $this->getRecipients($webformSettings);
+    $module = 'os2forms_webform_submission_log';
+    $key = 'submission_log_notification';
+    $langCode = $this->languageManager->getCurrentLanguage()->getId();
+
+    $params['title'] = 'Webform submission error - Form name: '. $webform->label() . ', Id: ' . $webformSubmission->serial();
+    $params['from'] = $this->configFactory->get('system.site')->get('mail');
+    $params['message'] = $this->createMessage($webform, $webformSubmission, $context);
+
+    foreach ($recipients as $to) {
+      $result = $this->mailManager->mail($module, $key, $to, $langCode, $params);
+      if ($result['result'] !== true) {
+        $logger = $this->getLogger($module);
+        $logger->error('There was a problem sending your email notification');
+      }
+    }
+  }
+
+  /**
+   * Get recipients.
+   *
+   * @param $webformSettings
+   *
+   * @return array[]
+   */
+  private function getRecipients($webformSettings): array {
+    return array_filter(array_map('trim', explode(PHP_EOL, $webformSettings['emails'])));
+  }
+
+  /**
+   * Create the message to send.
+   *
+   * @param $webform
+   *   The related webform.
+   * @param $webformSubmission
+   *   The related webform submission.
+   * @param $context
+   *   The related context.
+   *
+   * @return string
+   *   The created message.
+   */
+  private function createMessage($webform, $webformSubmission, $context): string {
+    $referenceUrl = $this->urlGenerator->generateFromRoute(
+      'entity.webform_submission.log',
+      [
+        'webform' => $webform->id(),
+        'webform_submission' => $webformSubmission->id()
+      ],
+      ['absolute' => TRUE]
+    );
```

Comment:
`UrlGeneratorInterface ::generateFromRoute` should not be used (cf. https://github.com/drupal/drupal/blob/10.1.x/core/lib/Drupal/Core/Routing/UrlGeneratorInterface.php#L77-L81).

The best way to generate an absolute url (without leaking metadata) may be

```php
Url::fromRoute(…)->setAbsolute()->toString(TRUE)->getGeneratedUrl();
```

Reply:
Fixed